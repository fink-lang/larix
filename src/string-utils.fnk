unindent_text = fn text, ind:
  [first, ...rest] = text.split rx/\n/g

  [...lines] = pipe rest:
    map line:
      line.slice ind

  [first, ...lines].join '\n'
