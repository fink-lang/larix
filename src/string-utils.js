import {curr_loc, next_loc} from '@fink/prattler';


export const unindent_text = (text, ind)=> (
  text
    .split(/\n/g)
    .map((line, idx)=> line.slice(idx > 0 ? ind : 0))
    .join('\n')
);


export const strip_block = (strings, ...parts)=> {
  const [, ...lines] = String.raw({raw: strings}, ...parts).split('\n');
  const ind = lines[0].search(/[^ ]/);
  const str = lines.map((lne)=> lne.slice(ind)).join('\n');
  return str;
};

