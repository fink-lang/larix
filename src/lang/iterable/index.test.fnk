{describe, it, eq, snapshot} = import '../../testing/jest'

{parse_expr} = import '../../'


describe:: 'fold item, accu: ...', fn:

  it:: 'parses', fn:
    parse_expr(`
      fold item, accu=0:
        item + accu
    `) eq snapshot


describe:: 'unfold item, accu: ...', fn:

  it:: 'parses', fn:
    parse_expr(`
      unfold curr=start, accu=0:
        (start + accu, accu + 1)
    `) eq snapshot


  it:: 'parses no accu', fn:
    parse_expr(`
      unfold curr=start:
        start + accu
    `) eq snapshot


  it:: 'parses no args', fn:
    parse_expr(`
      unfold:
        foo
    `) eq snapshot


describe:: 'map item: ...', fn:

  it:: 'parses', fn:
    parse_expr(`
      map item:
        item * 2
    `) eq snapshot


describe:: 'flat_map item: ...', fn:

  it:: 'parses', fn:
    parse_expr(`
      flat_map item:
        [item, item * 2]
    `) eq snapshot


describe:: 'filter item: ...', fn:

  it:: 'parses', fn:
    parse_expr(`
      filter item:
        item % 2 == 0
    `) eq snapshot


describe:: 'while item: ...', fn:

  it:: 'parses', fn:
    parse_expr(`
      while item:
        item % 2 == 0
    `) eq snapshot


describe:: 'find item: ...', fn:

  it:: 'parses', fn:
    parse_expr(`
      find item:
        item % 2 == 0
    `) eq snapshot

