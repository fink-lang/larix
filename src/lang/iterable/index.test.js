import {parse_expr} from '../../';
import {strip_block} from '../../string-utils';


describe('fold item, accu: ...', ()=> {

  it('parses', ()=> {
    expect(
      parse_expr(strip_block`
        fold item, accu=0:
          item + accu
      `)
    ).toMatchSnapshot();
  });
});


describe('unfold item, accu: ...', ()=> {

  it('parses', ()=> {
    expect(
      parse_expr(strip_block`
        unfold curr=start, accu=0:
          (start + accu, accu + 1)
      `)
    ).toMatchSnapshot();
  });


  it('parses no accu', ()=> {
    expect(
      parse_expr(strip_block`
        unfold curr=start:
          (start + accu, accu + 1)
      `)
    ).toMatchSnapshot();
  });


  it('parses no args', ()=> {
    expect(
      parse_expr(strip_block`
        unfold:
          foo
      `)
    ).toMatchSnapshot();
  });
});


describe('map item: ...', ()=> {

  it('parses', ()=> {
    expect(
      parse_expr(strip_block`
        map item:
          item * 2
      `)
    ).toMatchSnapshot();
  });
});


describe('flat_map item: ...', ()=> {

  it('parses', ()=> {
    expect(
      parse_expr(strip_block`
        flat_map item:
          [item, item * 2]
      `)
    ).toMatchSnapshot();
  });
});


describe('filter item: ...', ()=> {

  it('parses', ()=> {
    expect(
      parse_expr(strip_block`
        filter item:
          item % 2 == 0
      `)
    ).toMatchSnapshot();
  });
});
