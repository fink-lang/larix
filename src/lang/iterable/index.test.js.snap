// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter item: ... parses 1`] = `
block filter (1:0-2:15)
  other (1:7-1:11) item
  :
  comp == (2:2-2:15)
    arithm % (2:2-2:10)
      other (2:2-2:6) item
      other (2:9-2:10) 2
    other (2:14-2:15) 0
`;

exports[`flat_map item: ... parses 1`] = `
block flat_map (1:0-2:18)
  other (1:9-1:13) item
  :
  array (2:2-2:18)
    other (2:3-2:7) item
    arithm * (2:9-2:17)
      other (2:9-2:13) item
      other (2:16-2:17) 2
`;

exports[`fold item, accu: ... parses 1`] = `
block fold (1:0-2:13)
  other (1:5-1:9) item
  assign = (1:11-1:17)
    other (1:11-1:15) accu
    other (1:16-1:17) 0
  :
  arithm + (2:2-2:13)
    other (2:2-2:6) item
    other (2:9-2:13) accu
`;

exports[`map item: ... parses 1`] = `
block map (1:0-2:10)
  other (1:4-1:8) item
  :
  arithm * (2:2-2:10)
    other (2:2-2:6) item
    other (2:9-2:10) 2
`;

exports[`unfold item, accu: ... parses 1`] = `
block unfold (1:0-2:26)
  assign = (1:7-1:17)
    other (1:7-1:11) curr
    other (1:12-1:17) start
  assign = (1:19-1:25)
    other (1:19-1:23) accu
    other (1:24-1:25) 0
  :
  group (2:2-2:26)
    arithm + (2:3-2:15)
      other (2:3-2:8) start
      other (2:11-2:15) accu
    arithm + (2:17-2:25)
      other (2:17-2:21) accu
      other (2:24-2:25) 1
`;
